src/serial/dma.rs

```diff
@@ -0,0 +1,711 @@
+use core::{marker::PhantomData, mem::transmute, ops::Deref};
+
+use super::{Instance, RegisterBlockImpl, Serial};
+use crate::dma::{
+    config::DmaConfig,
+    traits::{Channel, DMASet, DmaFlagExt, PeriAddress, Stream, StreamISR},
+    ChannelX, MemoryToPeripheral, PeripheralToMemory, Transfer,
+};
+use crate::ReadFlags;
+
+use nb;
+
+#[non_exhaustive]
+pub enum Error {
+    SerialError(super::Error),
+    TransferError,
+}
+
+/// Tag for TX/RX channel that a corresponding channel should not be used in DMA mode
+#[non_exhaustive]
+pub struct NoDMA;
+
+/// Callback type to notify user code of completion serial transfers
+pub type SerialCompleteCallback = fn(Result<(), Error>);
+
+pub trait SerialWriteDMA {
+    /// Writes `bytes` to the serial interface in non-blocking mode
+    ///
+    /// # Arguments
+    /// * `bytes` - byte slice that need to send
+    /// * `callback` - callback that will be called on completion
+    ///
+    /// # Safety
+    /// This function relies on supplied slice `bytes` until `callback` called. So the slice must live until that moment.
+    ///
+    /// # Warning
+    /// `callback` may be called before function returns value. It happens on errors in preparation stages.
+    unsafe fn write_dma(
+        &mut self,
+        bytes: &[u8],
+        callback: Option<SerialCompleteCallback>,
+    ) -> nb::Result<(), super::Error>;
+}
+
+pub trait SerialReadDMA {
+    /// Reads bytes from the serial interface in non-blocking mode and writes these bytes in `buf`
+    ///
+    /// # Arguments
+    /// * `buf` - byte slice where received bytes will be written
+    /// * `callback` - callback that will be called on completion
+    ///
+    /// # Safety
+    /// This function relies on supplied slice `buf` until `callback` called. So the slice must live until that moment.
+    ///
+    /// # Warning
+    /// `callback` may be called before function returns value. It happens on errors in preparation stages.
+    unsafe fn read_dma(
+        &mut self,
+        buf: &mut [u8],
+        callback: Option<SerialCompleteCallback>,
+    ) -> nb::Result<(), super::Error>;
+}
+
+pub trait SerialWriteReadDMA {
```

Comment:
What use-case of WriteRead in with Serial?

Reply:
To be honest, I don't have a current use case for this. I was basically following the existing implementation of I2C and made it to similarly work for serial.

However, I do think there can be use case for such communication. For example, when communicating with another chip connected over serial, the WriteRead can be used to support the paradigm of send query and read response from the external chip.

Reply:
I2C is synchronous interface. UART is asynchronous.  This paradigm will not work properly without synchronization.

Please, remove this part from PR.

Reply:
Thanks. `SerialWriteReadDMA` is removed.